streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1418,16 +1555,119 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
+    ZStream.managed(self.toQueue()).flatMap { queue =>
+      ZStream
+        .unfoldM(()) { _ =>
+          queue.take.timeout(d).flatMap {
+            case Some(a) => ZIO.succeed(Some((a, ())))
+            case None    => ZIO.interrupt // upstream was too slow
+          }
+        }
+        .unTake
+    }
+
   /**
    * Converts the stream to a managed queue. After managed queue is used, the
    * queue will never again produce values and should be discarded.
    */
-  final def toQueue[E1 >: E, A1 >: A](capacity: Int = 2): ZManaged[R, E1, Queue[Take[E1, A1]]] =
+  final def toQueue[E1 >: E, A1 >: A](capacity: Int = 2): ZManaged[R, Nothing, Queue[Take[E1, A1]]] =
     for {
       queue <- ZManaged.make(Queue.bounded[Take[E1, A1]](capacity))(_.shutdown)
       _     <- self.intoManaged(queue).fork
     } yield queue
 
+  /**
+   * Converts the stream to a managed list of queues. Every value will be replicated to every queue with the
+   * slowest queue being allowed to buffer maximumLag elements before the driver is backpressured.
+   * The downstream queues will be provided with elements in the same order they are returned, so
+   * the fastest queue might have seen up to (maximumLag + 1) elements more than the slowest queue if it
+   * has a lower index than the slowest queue.
+   * During the finalizer the driver will wait for all queues to shutdown in order to signal completion.
+   * After managed queue is used, the queue will never again produce values and should be discarded.
+   */
+  final def toQueues[E1 >: E, A1 >: A](n: Int, maximumLag: Int): ZManaged[R, Nothing, List[Queue[Take[E1, A1]]]] = {
+    val decider = ZIO.succeed((_: Int) => true)
+    toQueuesBalanced(n, maximumLag, _ => decider)
+  }
+
+  /**
+   * More powerful version of `ZStream#toQueues`. Allows to provide a function that determines what
+   * queues should receive which elements.
+   */
+  final def toQueuesBalanced[E1 >: E, A1 >: A](
+    n: Int,
+    maximumLag: Int,
+    decide: A => UIO[Int => Boolean]
+  ): ZManaged[R, Nothing, List[Queue[Take[E1, A1]]]] =
+    Promise.make[Nothing, A => UIO[Int => Boolean]].toManaged_.flatMap { prom =>
+      toQueuesBalanced0[E1, A1](maximumLag, (a: A) => prom.await.flatMap(_(a)), _ => ZIO.unit).flatMap { next =>
+        (ZIO.collectAll(List.fill(n)(next.map(_._2))) <* prom.succeed(decide)).toManaged_
+      }
+    }
+
+  /**
+   * More powerful version of `ZStream#toQueuesBalanced`. This returns a function that will produce
+   * new queues and corresponding indices.
+   * You can also provide a function that will be executed after the final events are enqueued in all queues
+   * but before the queues are actually shutdown.
+   * Shutdown of the queues is handled by the driver.
+   */
+  final def toQueuesBalanced0[E1 >: E, A1 >: A](
+    maximumLag: Int,
+    decide: A => UIO[Int => Boolean],
+    done: Take[E1, Nothing] => UIO[_]
+  ): ZManaged[R, Nothing, UIO[(Int, Queue[Take[E1, A1]])]] =
```

Comment:
`distributedDynamicWith`? As this version will allow creating additional queues dynamically